changes/changes.go

```diff
@@ -0,0 +1,602 @@
+}
+
+// ChangeReportDeltaComponent is a subcomponent that can hold
+// delta information between left and right
+type ChangeReportDeltaComponent struct {
+	ChangeReportComponent
+	Title string      `json:"title,omitempty"`
+	Delta interface{} `json:"delta"`
+}
+
+// StatsChangeComponent represents the stats change report
+type StatsChangeComponent struct {
+	Summary *ChangeReportDeltaComponent   `json:"summary"`
+	Columns []*ChangeReportDeltaComponent `json:"columns"`
+}
+
+// ChangeReportResponse is the result of a call to changereport
+type ChangeReportResponse struct {
+	VersionInfo *ChangeReportComponent `json:"version_info,omitempty"`
+	Commit      *ChangeReportComponent `json:"commit,omitempty"`
+	Meta        *ChangeReportComponent `json:"meta,omitempty"`
+	Readme      *ChangeReportComponent `json:"readme,omitempty"`
+	Structure   *ChangeReportComponent `json:"structure,omitempty"`
+	Transform   *ChangeReportComponent `json:"transform,omitempty"`
+	Stats       *StatsChangeComponent  `json:"stats,omitempty"`
+}
+
+// StatsChangeSummaryFields represents the stats summary
+type StatsChangeSummaryFields struct {
+	Entries int `json:"entries"`
+	Columns int `json:"columns"`
+	// NullValues int `json:"nullValues"`
+	TotalSize int `json:"totalSize"`
+}
+
+// EmptyObject is used mostly as a placeholder in palces where it is required
+// that a key is present in the response even if empty and not be nil
+type EmptyObject map[string]interface{}
+
+// Service can generate a change report between two datasets
+type Service struct {
+	loader dsref.Loader
+	stats  *stats.Service
+}
+
+// New allocates a Change service
+func New(loader dsref.Loader, stats *stats.Service) *Service {
+	return &Service{
+		loader: loader,
+		stats:  stats,
+	}
+}
+
+func (svc *Service) parseColumns(colItems *tabular.Columns, data *dataset.Dataset) (interface{}, error) {
+	var sErr error
+	if data.Structure != nil {
+		*colItems, _, sErr = tabular.ColumnsFromJSONSchema(data.Structure.Schema)
+		if sErr != nil {
+			return nil, sErr
+		}
+		return StatsChangeSummaryFields{
+			Entries:   data.Structure.Entries,
+			Columns:   len(*colItems),
+			TotalSize: data.Structure.Length,
+		}, nil
+	}
+	return EmptyObject{}, nil
+}
+
+// maybeLoadStats attempts to load stats if not alredy present
+// errors out if it fails as stats are required and some datasets might not yet have
+// a stats component attached to it
+func (svc *Service) maybeLoadStats(ctx context.Context, ds *dataset.Dataset) error {
+	if ds.Stats != nil {
+		return nil
+	}
+	var statsErr error
+	ds.Stats, statsErr = svc.stats.Stats(ctx, ds)
+	if statsErr != nil {
+		return qerr.New(statsErr, "missing stats components")
+	}
+	return nil
+}
+
+// parseStats uses json serializing > deserializing to easily parse the stats
+// interface as we have little type safety in the dataset.stats component right now
+func (svc *Service) parseStats(ds *dataset.Dataset) ([]EmptyObject, error) {
```

Comment:
into it. We should get on this ASAP

Reply:
What is the shape of `Stats`? It looks to be declared as `interface{}`, but is it always actually a list of json like objects? In other words `[]map[string]interface{}`?

If so, it is probably worth adding a method to `base/fill` that operates on a list of structs. I'm thinking `fill.StructList` or something, that just loops over the input and calls `fill.Struct` for each element. Seems like a good fit, this is exactly what `base/fill` was designed to do.

Reply:
This is mainly a shortcut to work around type casting issues between `interface{}` and `[]something{}` hence the double serialization. In this specific usecase the top-level is indeed a `[]map[string]interface{}` however it is not *always* the case as far as I understood it from the dataset package especially in the realm of JSON datasets. 

Will look into it if it makes sense doing now or simply working with this untill we introduce the propper typing in `Stats`